Preserve bond type and order in all `Modeller` operations

nicoladmSB opened the pull request:
This is not linked to an open issue (as far as I can tell), but it falls in the same category as #4112 (issue), #4113 (PR), completing the work of the latter. 

----

When generating a new topology for a `Modeller` as a result of any operation, bonds that had `type` and `order` attributes in the original topology should retain them, instead of implicitly setting them to `None`.

This PR

- ensures that  `delete()` (therefore implicitly `deleteWater()`), `addSolvent()`, `addHydrogens()` and `addExtraParticles()` behave in the desirable way, while #4112 did it for `add()` (and implicitly for `addMembrane()`).

- adds regression tests either as new tests method, or as assertions in existing test methods

peastman commented:
Thanks!  The change looks good.

I think the tests could be simplified.  Rather than modifying a bunch of existing tests, how about adding a single new test that runs through a series of manipulations?

- Create a Topology with a small molecule that includes bond types.
- Call `addSolvent()` to add some water.
- Call `convertWater()` to convert it to TIP4P.
- Call `addExtraParticles()` to add the virtual sites.
- Call `delete()` to remove some of the water.
- Verify that the bond information for the small molecule has survived all those calls intact.

nicoladmSB commented:
Thanks for the review!

I've reverted the changes to the test module, and replaced them with a single test that does more or less what you recommended. I've just added a couple of steps to it, to ensure that `add()`, `addHydrogens()` and `deleteWater()` are also exercised.

peastman commented:
Looks good.  That's a really nice test case.
